File: examples/backend/src/test/java/com/example/agentsdlc/architecture/RatchetGuardTest.java
package com.example.agentsdlc.architecture;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.junit.jupiter.api.Test;

/**
 * Gate 9 — the backend half of the ratchet guard: a test whose only job is to read the
 * LIVE configs and fail if a floor was lowered, a threshold deleted, or a freeze store
 * removed. `fast-repo-hygiene`'s render-floors.sh diff (design.md section 7.5) catches a
 * hand-edit at pull-request time; this test is the one that survives even if that CI job
 * is ever deleted or bypassed, because it runs inside `mvn verify` itself.
 *
 * Never edit this test, floors.yml, or a FLOORS:BEGIN/END block to make a red run go
 * green — see docs/QUALITY-GATES.md's ratchet policy. That is an escalation, not a fix.
 */
class RatchetGuardTest {

  private static final Pattern FLOOR_ENTRY =
      Pattern.compile("^ {2}([a-z0-9_.]+):\\s*\\{\\s*value:\\s*([^,}]+)", Pattern.MULTILINE);

  private String floorsYml() throws IOException {
    Path repoRoot = findRepoRoot();
    return Files.readString(repoRoot.resolve("floors.yml"));
  }

  private String pomXml() throws IOException {
    Path repoRoot = findRepoRoot();
    return Files.readString(repoRoot.resolve("examples/backend/pom.xml"));
  }

  // Walks up from the working directory (Maven always runs `mvn` from backend/) to find
  // the repo root that holds floors.yml — no hardcoded absolute path, so this test works
  // from any checkout location.
  private Path findRepoRoot() {
    Path dir = Paths.get("").toAbsolutePath();
    while (dir != null) {
      if (Files.exists(dir.resolve("floors.yml"))) {
        return dir;
      }
      dir = dir.getParent();
    }
    throw new IllegalStateException("could not find floors.yml above " + Paths.get("").toAbsolutePath());
  }

  private String floorValue(String yml, String key) {
    Matcher m = FLOOR_ENTRY.matcher(yml);
    while (m.find()) {
      if (m.group(1).equals(key)) {
        return m.group(2).trim();
      }
    }
    // Calibrated (multi-line block) form: "  key:\n    value: X".
    Matcher block = Pattern.compile(
            "^ {2}" + Pattern.quote(key) + ":\\s*\\n\\s*value:\\s*([^\\n]+)",
            Pattern.MULTILINE)
        .matcher(yml);
    if (block.find()) {
      return block.group(1).trim();
    }
    throw new IllegalStateException("floors.yml has no entry for " + key);
  }

  @Test
  void jacocoRatchetAgreesWithFloorsYml() throws IOException {
    String yml = floorsYml();
    String pom = pomXml();
    String line = floorValue(yml, "backend.coverage.line");
    String branch = floorValue(yml, "backend.coverage.branch");

    if (line.equals("unset") || branch.equals("unset")) {
      assertThat(pom)
          .as("floors.yml has an unset backend coverage floor; pom.xml's jacoco:check must be skipped, not silently thresholded")
          .contains("<skip>true</skip>");
    } else {
      assertThat(pom)
          .as("floors.yml has calibrated backend coverage floors; pom.xml's jacoco:check must be armed, not skipped")
          .contains("<skip>false</skip>")
          .contains("<minimum>" + line + "</minimum>")
          .contains("<minimum>" + branch + "</minimum>");
    }
  }

  @Test
  void pitRatchetAgreesWithFloorsYml() throws IOException {
    String yml = floorsYml();
    String pom = pomXml();
    String score = floorValue(yml, "backend.mutation.score");

    if (score.equals("unset")) {
      assertThat(pom)
          .as("floors.yml has an unset mutation floor; pom.xml's pit.mutationThreshold must read as the tool's own uncalibrated value")
          .contains("<pit.mutationThreshold>0</pit.mutationThreshold>");
    } else {
      int expectedPercent = (int) Math.round(Double.parseDouble(score) * 100);
      assertThat(pom)
          .as("floors.yml has a calibrated mutation floor; pom.xml's pit.mutationThreshold must match it")
          .contains("<pit.mutationThreshold>" + expectedPercent + "</pit.mutationThreshold>");
    }
  }

  @Test
  void theArchitectureFreezeStoreHasNotBeenDeleted() {
    Path repoRoot = findRepoRoot();
    Path store = repoRoot.resolve("examples/backend/archunit_store");
    assertThat(Files.isDirectory(store))
        .as("the archunit_store/ freeze store must exist — see LayeredArchitectureTest and docs/QUALITY-GATES.md's ratchet policy on never removing a freeze store")
        .isTrue();
  }

  @Test
  void theMigrationDirectoryIsNeverEmptyAndCarriesNoDuplicateVersions() throws IOException {
    Path repoRoot = findRepoRoot();
    Path migrations = repoRoot.resolve("examples/backend/src/main/resources/db/migration");
    assertThat(Files.isDirectory(migrations)).isTrue();

    try (var files = Files.list(migrations)) {
      var versions = files
          .map(p -> p.getFileName().toString())
          .filter(name -> name.matches("V\\d+(\\.\\d+)*__.*\\.sql"))
          .map(name -> name.split("__")[0])
          .toList();
      assertThat(versions).as("no migrations found").isNotEmpty();
      assertThat(versions).as("duplicate Flyway version numbers").doesNotHaveDuplicates();
    }
  }
}

File: examples/frontend/src/ratchetGuard.test.ts
// Gate 9 — the frontend half of the ratchet guard: reads the LIVE configs and fails if a
// floor was lowered, a threshold deleted, or an exclude widened. `fast-repo-hygiene`'s
// render-floors.sh diff (design.md section 7.5) catches a hand-edit at pull-request
// time; this test is the one that survives even if that CI job is ever deleted or
// bypassed, because it runs inside `npm run test:coverage` itself.
//
// Never edit this test, floors.yml, or a FLOORS:BEGIN/END block to make a red run go
// green — see docs/QUALITY-GATES.md's ratchet policy. That is an escalation, not a fix.
import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

function findRepoRoot(): string {
  let dir = path.resolve(__dirname, "..", "..");
  while (dir !== path.parse(dir).root) {
    if (
      (() => {
        try {
          readFileSync(path.join(dir, "floors.yml"));
          return true;
        } catch {
          return false;
        }
      })()
    ) {
      return dir;
    }
    dir = path.dirname(dir);
  }
  throw new Error("could not find floors.yml above " + __dirname);
}

function floorValue(yml: string, key: string): string {
  const escaped = key.replace(/\./g, "\\.");
  const flow = yml.match(new RegExp(`^ {2}${escaped}:\\s*\\{\\s*value:\\s*([^,}]+)`, "m"));
  if (flow) return flow[1].trim();
  const block = yml.match(new RegExp(`^ {2}${escaped}:\\s*\\n\\s*value:\\s*([^\\n]+)`, "m"));
  if (block) return block[1].trim();
  throw new Error(`floors.yml has no entry for ${key}`);
}

const ROOT = findRepoRoot();
const floorsYml = readFileSync(path.join(ROOT, "floors.yml"), "utf8");
const vitestConfig = readFileSync(path.join(ROOT, "examples/frontend/vitest.config.js"), "utf8");
const strykerConfig = readFileSync(path.join(ROOT, "examples/frontend/stryker.config.mjs"), "utf8");

describe("ratchet guard", () => {
  it("vitest coverage thresholds agree with floors.yml", () => {
    const keys = [
      "frontend.coverage.statements",
      "frontend.coverage.branches",
      "frontend.coverage.functions",
      "frontend.coverage.lines",
    ];
    const values = keys.map((k) => floorValue(floorsYml, k));
    const anyUnset = values.some((v) => v === "unset");

    if (anyUnset) {
      expect(
        vitestConfig,
        "floors.yml has an unset frontend coverage floor; vitest.config.js's thresholds must be the empty object, not silently populated",
      ).toMatch(/thresholds:\s*\{\}/);
    } else {
      const [stmts, branches, funcs, lines] = values.map((v) =>
        Math.round(Number(v) * 100),
      );
      expect(
        vitestConfig,
        "floors.yml has calibrated frontend coverage floors; vitest.config.js's thresholds must match them",
      ).toContain(
        `thresholds: { statements: ${stmts}, branches: ${branches}, functions: ${funcs}, lines: ${lines} }`,
      );
    }
  });

  it("Stryker's mutation threshold agrees with floors.yml", () => {
    const score = floorValue(floorsYml, "frontend.mutation.score");
    if (score === "unset") {
      expect(
        strykerConfig,
        "floors.yml has an unset mutation floor; stryker.config.mjs's break must stay null, Stryker's own documented 'never break the build'",
      ).toMatch(/break:\s*null/);
    } else {
      const expected = Math.round(Number(score) * 100);
      expect(strykerConfig).toContain(`break: ${expected}`);
    }
  });

  it("the design-system guardrail script has not been removed", () => {
    // A removed check is indistinguishable from a passing one unless something asserts
    // it still exists — same shape as the freeze-store-deletion guard on the backend.
    const scriptPath = path.join(ROOT, "examples/frontend/scripts/check-design-system.mjs");
    expect(() => readFileSync(scriptPath)).not.toThrow();
  });
});
